addons/website_sale/models/product_product.py

```diff
@@ -137,7 +137,12 @@ def _website_show_quick_add(self):
 
     def _is_add_to_cart_allowed(self):
         self.ensure_one()
-        return self.user_has_groups('base.group_system') or (self.active and self.sale_ok and self.website_published)
+        website = self.env['website'].get_current_website()
+        is_allowed = (self.user_has_groups('base.group_system')
+                      or (self.active and self.sale_ok and self.website_published
+                          and not (website.ecommerce_access == 'logged'
+                                   and self.env.user._is_public())))
```

Comment:
A bit confusing, maybe we could do smth like that to make it more readable? Feel free to find more suitable names:) 
```suggestion
        has_shop_access = not (website.ecommerce_access == 'logged' and self.env.user._is_public())
        is_product_ready = self.active and self.sale_ok and self.website_published
        return self.user_has_groups('base.group_system') or is_product_ready and has_shop_access
```